### PR TITLE
SDK integration (program inputs)

### PR DESCRIPTION
**Pull request by moodlezoup**
Uses new Jolt SDK to feed public inputs to guest program.
Also fixes some issues around load and store instructions. 
TODO: Verifier currently reconstructs the full v_init memory polynomial when computing v_init(r). This is fine for programs with small memory usage, but this opening should take O(|bytecode| + |inputs|) instead of O(|memory|)

### CODE REVIEW

**Review by sragss — APPROVED** [review]
LGTM

File: common/src/constants.rs [hunk 1]
```diff
@@ -1,7 +1,29 @@
...
+pub const fn memory_address_to_witness_index(address: u64) -> usize {
+    (address + RAM_WITNESS_OFFSET - RAM_START_ADDRESS) as usize
+}
+pub const fn witness_index_to_memory_address(index: usize) -> u64 {
+    index as u64 + RAM_START_ADDRESS - RAM_WITNESS_OFFSET
+}
+// Layout of the witness (where || denotes concatenation):
+//     registers || inputs || outputs || panic || padding || RAM
+// Layout of VM memory:
+//     peripheral devices || inputs || outputs || panic || padding || RAM
```
[the remark this answers is not in the document]
Comment (moodlezoup) [reply common/src/constants.rs:27]: They are treated as a peripheral device, but peripheral devices have memory addresses too. So we just make sure the device's address space is contiguous with the RAM address space

File: examples/sha3-ex/guest/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+cargo-features = ["per-package-target"]
+
+[package]
+name = "sha3-guest"
+version = "0.1.0"
+edition = "2021"
+
+forced-target = "riscv32i-unknown-none-elf"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
[the remark this answers is not in the document]
Comment (moodlezoup) [reply examples/sha3-ex/guest/Cargo.toml:10]: done in #205

File: examples/sha2-ex/guest/linker.ld [hunk 3]
```diff
@@ -1,5 +1,5 @@
 MEMORY {
-  program (rwx) : ORIGIN = 0x80000000, LENGTH = 2 * 1024 * 1024
+  program (rwx) : ORIGIN = 0x80000000, LENGTH = 10 * 1024 * 1024
```
[the remark this answers is not in the document]
Comment (moodlezoup) [reply examples/sha2-ex/guest/linker.ld:2]: @ncitron is this effectively max memory size?

File: common/src/rv_trace.rs [hunk 4]
```diff
@@ -626,4 +640,56 @@ impl RV32IM {
             | RV32IM::UNIMPL => unimplemented!(),
         }
     }
-}
\ No newline at end of file
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
```
[the remark this answers is not in the document]
Comment (moodlezoup) [reply common/src/rv_trace.rs:649]: done

**Pull request merged**